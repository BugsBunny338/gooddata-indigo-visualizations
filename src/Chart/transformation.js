/* eslint no-underscore-dangle: 0 */
import cloneDeep from 'lodash/cloneDeep';
import get from 'lodash/get';
import map from 'lodash/map';
import sortBy from 'lodash/sortBy';
import compact from 'lodash/compact';
import uniqBy from 'lodash/uniqBy';
import times from 'lodash/times';
import constant from 'lodash/constant';
import filter from 'lodash/filter';

// default format used when creating metric
export const DEFAULT_FORMAT = '#,##0.00';

/**
 * Splits metrics to two separate groups.
 *
 * One group contains all the metric headers, the other contains everything else.
 */
export function _splitHeaders(headers) {
    const metrics = [];
    const newHeaders = [];

    headers.forEach((header, i) => {
        if (header.type === 'metric') {
            metrics.push({
                index: i,
                header
            });
        } else {
            newHeaders.push(header);
        }
    });

    return {
        headers: newHeaders,
        metrics
    };
}

export function getMetricNamesValuesHeaderItems(headers, metrics) {
    // Need to transform the headers to match the rawData transformation
    return [{
        id: 'metricNames',
        title: 'Metric',
        type: 'attrLabel',
        uri: '/metricGroup',
        metrics
    }, {
        id: 'metricValues',
        type: 'metric',
        uri: '/metricValues',

        // For axis formatting, use first metric's formatting.
        // For individual series tooltips and labels, formats are
        // handled separately in metricNames
        format: get(metrics, '0.header.format', DEFAULT_FORMAT),

        // use first metric title as default
        title: get(metrics, '0.header.title', '')
    }];
}

function addMetricNamesValues(headers, metrics) {
    return {
        headers: headers.concat(getMetricNamesValuesHeaderItems(headers, metrics)),
        metrics
    };
}

export function enrichHeaders(_headers) {
    const { headers, metrics } = _splitHeaders(_headers);

    return addMetricNamesValues(headers, metrics);
}

export function _transposeData(headers, metrics, rawData) {
    const data = [];

    for (let mi = 0; mi < metrics.length; mi++) {
        for (let ri = 0; ri < rawData.length; ri++) {
            const row = [];

            for (let ci = 0; ci < headers.length; ci++) {
                if (headers[ci].type === 'attrLabel') {
                    const value = rawData[ri][ci];
                    row.push({
                        id: value,
                        value
                    });
                }
            }

            const metric = metrics[mi];
            row.push({
                id: `metric-${mi}`, // let each metric have unique id
                value: metric.header.title
            });

            const metricValue = rawData[ri][metric.index];
            const parsedValue = parseFloat(metricValue);
            const y = isNaN(parsedValue) ? null : parsedValue;

            row.push({
                format: metric.header.format || DEFAULT_FORMAT,
                y,
                marker: { enabled: y !== null }
            });

            data.push(row);
        }
    }

    return data;
}

/**
 * For multiple metrics data transform |A1|A2|M1|M2|M3| format
 * into a |A1|A2|metricNames|metricValues| format where all the metrics
 * are put underneath itself and the metricNames acts as a report-local
 * artificial attribute with individual metric names are its values
 *
 * Note: output is wrapped into an object having `id` and `value` keys
 * for the purposes of determining unique elements
 *
 * Example:
 * <pre>
 *
 *   Input
 *
 *     A1  A2    M1     M2
 *    -----------------------
 *      u   d   1111   2333
 *      u   e   1112   2334
 *      v   o   1121   2345
 *      v   p   1122   2346
 *      v   q   1123   2347
 *
 *   Output
 *
 *     A1  A2  metricNames metricValues
 *    ----------------------------------
 *      u   d       M1         1111    // first metric
 *      u   e       M1         1112
 *      v   o       M1         1121
 *      v   p       M1         1122
 *      v   q       M1         1123
 *      u   d       M2         2333    // second metric
 *      u   e       M2         2334
 *      v   o       M2         2345
 *      v   p       M2         2346
 *      v   q       M2         2347
 *
 * </pre>
 **/
export function _transformMetrics(data) {
    // don't modify original data structure
    const dataCopy = cloneDeep(data);

    const { headers, metrics } = enrichHeaders(dataCopy.headers);

    if (metrics.length < 1) return dataCopy;

    const transposedData = _transposeData(dataCopy.headers, metrics, dataCopy.rawData);

    dataCopy.rawData = transposedData;
    dataCopy.headers = headers;
    return dataCopy;
}

export function transformData(data) {
    if (!data || data.isLoading) return data;

    return _transformMetrics(data);
}

// get unique elements for given index regarding their interal id
export function _getElements(data, index) {
    return uniqBy(map(data.rawData, index), 'id');
}

// get series data
export function _getSeries(data, seriesNames, categories, indices) {
    const seriesData = seriesNames.reduce((acc, series) => {
        acc[series.id] = times( // eslint-disable-line no-param-reassign
            categories.length,
            constant(null)
        );
        return acc;
    }, {});

    // prepare data points
    if (indices.series !== -1) {
        if (indices.category !== -1) {
            const categoriesIndex = categories.reduce((acc, category, idx) => {
                if (category) {
                    acc[category.id] = idx; // eslint-disable-line no-param-reassign
                }
                return acc;
            }, {});

            data.rawData.forEach((dataRow) => {
                const categoryItemIndex = categoriesIndex[dataRow[indices.category].id];
                seriesData[dataRow[indices.series].id][categoryItemIndex] = dataRow[indices.metric];
            });
        } else {
            data.rawData.forEach((dataRow) => {
                seriesData[dataRow[indices.series].id][0] = dataRow[indices.metric];
            });
        }
    }

    // return series in the same order as requested in seriesNames
    return seriesNames.map((name, index) => {
        return {
            name: name.value,
            legendIndex: index,
            data: seriesData[name.id]
        };
    });
}

function lighter(color, percent) {
    const t = percent < 0 ? 0 : 255;
    const p = Math.abs(percent);

    return Math.round((t - color) * p) + color;
}

function formatColor(red, green, blue) {
    return `rgb(${red},${green},${blue})`;
}

/**
 * Source:
 *     http://stackoverflow.com/questions/5560248/programmatically-lighten-or-darken-a-hex-color-or-rgb-and-blend-colors
 */
export function _getLighterColor(color, percent) {
    const f = color.split(',');
    const R = parseInt(f[0].slice(4), 10);
    const G = parseInt(f[1], 10);
    const B = parseInt(f[2], 10);

    return formatColor(
        lighter(R, percent),
        lighter(G, percent),
        lighter(B, percent)
    );
}

export function getColorPalette(data, palette) {
    const newPalette = cloneDeep(palette);

    filter(data.headers, header => header.type === 'metric')
        .forEach((metric, idx) => {
            if (metric.id && metric.id.match(/\.generated\.(filtered_)?pop\./)) {
                const color = _getLighterColor(newPalette[idx % newPalette.length], 0.6);
                newPalette.splice(idx, 0, color);
            }
        });

    return newPalette;
}

/**
 * Get highcharts-compatible charting data from the data which
 * already passed a metricNames+metricValues transformation
 * Needs configuration for categories/series and metric values
 *
 * Input
 * Example:
 * <pre>
 *
 *   Input
 *
 *     A  metricNames metricValues
 *    ------------------------------
 *     a       M1         1111    // first metric
 *     b       M1         1112
 *     c       M1         1121
 *     a       M2         2333    // second metric
 *     b       M2         2334
 *     c       M2         2345
 *
 *   Input configuration:
 *      sort: false,
 *      indices: {
 *          categories: 0,
 *          series: 1,
 *          metric: 2
 *      }
 *
 * </pre>
 *
 * Output
 * Example:
 *      {
 *          categories: ['a', 'b', 'c'],
 *          series: [
 *              {
 *                  name: 'M1',
 *                  data: [ { y: 1111 }, { y: 1112 }, { y: 1121 } ]
 *              },
 *              {
 *                  name: 'M2',
 *                  data: [ { y: 2333 }, { y: 2334 }, { y: 2345 } ]
 *              }
 *          ]
 *      }
 * <pre>
 *
 * </pre>
 **/
export function getChartData(data, configuration) {
    const indices = configuration.indices;

    const categories = _getElements(data, indices.category);
    const seriesNames = _getElements(data, indices.series);

    const seriesNamesSorted = configuration.sortSeries ? sortBy(seriesNames, 'value') : seriesNames;
    const seriesNamesCompact = compact(seriesNamesSorted);

    const seriesData = _getSeries(data, seriesNamesCompact, categories, indices);

    return {
        categories: map(categories, 'value'),
        series: seriesData
    };
}
