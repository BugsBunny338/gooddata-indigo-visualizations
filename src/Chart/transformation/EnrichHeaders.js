import { get } from 'lodash';
import { DEFAULT_FORMAT } from '../constants/metrics';

export function getMetricNamesValuesHeaderItems(headers, metrics) {
    // Need to transform the headers to match the rawData transformation
    return [{
        id: 'metricNames',
        title: 'Metric',
        type: 'attrLabel',
        uri: '/metricGroup',
        metrics
    }, {
        id: 'metricValues',
        type: 'metric',
        uri: '/metricValues',

        // For axis formatting, use first metric's formatting.
        // For individual series tooltips and labels, formats are
        // handled separately in metricNames
        format: get(metrics, '0.header.format', DEFAULT_FORMAT),

        // use first metric title as default
        title: get(metrics, '0.header.title', '')
    }];
}

function addMetricNamesValues(headers, metrics) {
    return {
        headers: [...headers, ...getMetricNamesValuesHeaderItems(headers, metrics)],
        metrics
    };
}

function splitHeaders(headers) {
    const metrics = [];
    const otherHeaders = [];

    headers.forEach((header, i) => {
        if (header.type === 'metric') {
            metrics.push({
                index: i,
                header
            });
        } else {
            otherHeaders.push(header);
        }
    });

    return {
        headers: otherHeaders,
        metrics
    };
}

// TODO: Cover all the module with tests
export function enrichHeaders(_headers) {
    const { headers, metrics } = splitHeaders(_headers);

    return addMetricNamesValues(headers, metrics);
}
