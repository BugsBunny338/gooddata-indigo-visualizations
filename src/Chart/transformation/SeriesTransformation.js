import { compact, uniqBy, map, times, constant, set, sortBy, get } from 'lodash';
import { enableDrillablePoints } from '../../utils/drilldownEventing';

// get unique elements for given index regarding their interal id
function getElements(data, index) {
    return uniqBy(map(data.rawData, index), 'id');
}

// get series data
function getSeries(data, seriesNames, categories, indices, drillableItems = []) {
    const seriesData = seriesNames.reduce((acc, series) => {
        return set(acc, series.id, times(categories.length, constant(null)));
    }, {});

    // prepare data points
    if (indices.series !== -1) {
        if (indices.category !== -1) {
            const categoriesIndex = categories.reduce((acc, category, idx) => {
                return category ? set(acc, category.id, idx) : acc;
            }, {});

            data.rawData.forEach((dataRow) => {
                const categoryItemIndex = categoriesIndex[dataRow[indices.category].id];
                const pointDrill = enableDrillablePoints(
                    drillableItems,
                    dataRow[indices.metric],
                    dataRow.slice(0, -1)
                );
                set(seriesData, [dataRow[indices.series].id, categoryItemIndex], pointDrill);
            });
        } else {
            data.rawData.forEach((dataRow) => {
                const pointDrill = enableDrillablePoints(
                    drillableItems,
                    dataRow[indices.metric],
                    dataRow.slice(0, -1)
                );
                set(seriesData, [dataRow[indices.series].id, 0], pointDrill);
            });
        }
    }

    // return series in the same order as requested in seriesNames
    return seriesNames.map((name, index) => {
        return {
            name: name.value,
            legendIndex: index,
            data: seriesData[name.id]
        };
    });
}

function setDrillableFlag(series) {
    return series.map((s) => {
        return {
            ...s,
            isDrillable: get(s, 'data.0.drilldown')
        };
    });
}

/**
 * Get highcharts-compatible charting data from the data which
 * already passed a metricNames+metricValues transformation
 * Needs configuration for categories/series and metric values
 *
 * Input
 * Example:
 * <pre>
 *
 *   Input
 *
 *     A  metricNames metricValues
 *    ------------------------------
 *     a       M1         1111    // first metric
 *     b       M1         1112
 *     c       M1         1121
 *     a       M2         2333    // second metric
 *     b       M2         2334
 *     c       M2         2345
 *
 *   Input configuration:
 *      sort: false,
 *      indices: {
 *          categories: 0,
 *          series: 1,
 *          metric: 2
 *      }
 *
 * </pre>
 *
 * Output
 * Example:
 *      {
 *          categories: ['a', 'b', 'c'],
 *          series: [
 *              {
 *                  name: 'M1',
 *                  data: [ { y: 1111 }, { y: 1112 }, { y: 1121 } ]
 *              },
 *              {
 *                  name: 'M2',
 *                  data: [ { y: 2333 }, { y: 2334 }, { y: 2345 } ]
 *              }
 *          ]
 *      }
 * <pre>
 *
 * </pre>
 **/
export function getChartData(data, configuration, drillableItems) {
    const indices = configuration.indices;

    const categories = getElements(data, indices.category);
    const seriesNames = getElements(data, indices.series);

    const seriesNamesSorted = configuration.sortSeries ? sortBy(seriesNames, 'value') : seriesNames;
    const seriesNamesCompact = compact(seriesNamesSorted);

    const series = getSeries(data, seriesNamesCompact, categories, indices, drillableItems);

    return {
        categories: map(categories, 'value'),
        series: setDrillableFlag(series)
    };
}
