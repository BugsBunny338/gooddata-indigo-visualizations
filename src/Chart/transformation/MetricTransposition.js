import { cloneDeep } from 'lodash';
import { enrichHeaders } from './EnrichHeaders';
import { parseValue } from '../../utils/common';

import { DEFAULT_FORMAT } from '../constants/metrics';

function invalidInput(data) {
    return !data || data.isLoading;
}

function transposeData(headers, metrics, rawData) {
    const data = [];

    for (let mi = 0; mi < metrics.length; mi++) {
        for (let ri = 0; ri < rawData.length; ri++) {
            const row = [];

            for (let ci = 0; ci < headers.length; ci++) {
                if (headers[ci].type === 'attrLabel') {
                    const element = rawData[ri][ci];

                    row.push({
                        id: element.id,
                        value: element.name,
                        header: {
                            id: headers[ci].id,
                            uri: headers[ci].uri
                        }
                    });
                }
            }

            const metric = metrics[mi];

            row.push({
                id: metric.header.id, // each metric must have unique id
                value: metric.header.title,
                header: {
                    id: metric.header.id ? metric.header.id : null,
                    uri: metric.header.uri ? metric.header.uri : null
                }
            });

            const metricValue = rawData[ri][metric.index];
            const y = parseValue(metricValue);
            row.push({
                format: metric.header.format || DEFAULT_FORMAT,
                y,
                marker: { enabled: y !== null }
            });

            data.push(row);
        }
    }

    return data;
}

export function transposeMetrics(data) {
    if (invalidInput(data)) {
        return data;
    }

    const dataCopy = cloneDeep(data);

    const { headers, metrics } = enrichHeaders(dataCopy.headers);

    if (metrics.length < 1) return dataCopy;

    return {
        ...dataCopy,
        headers,
        rawData: transposeData(dataCopy.headers, metrics, dataCopy.rawData)
    };
}
