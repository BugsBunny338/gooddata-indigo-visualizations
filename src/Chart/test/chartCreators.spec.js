import {
    propertiesToHeaders,
    getIndices,
    isMetricNamesInSeries,
    getLegendLayout,
    getCategoryAxisLabel,
    getMetricAxisLabel,
    showInPercent,
    generateTooltipFn,
    generatePieTooltipFn,
    getLineFamilyChartData,
    getPieFamilyChartData,
    getPieChartOptions
} from '../chartCreators';

import {
    DEFAULT_COLOR_PALETTE
} from '../transformation';

import { transposeMetrics } from '../transformation/MetricTransposition';

describe('chartCreators', () => {
    let config;
    let mockData;
    let mockHeaders;

    beforeEach(() => {
        config = {
            type: 'column',
            x: '/gdc/md/zro9kxjp2hejksfug8qemqwx6d92c940/obj/2577',
            y: '/metricValues',
            color: '/metricGroup',
            colorPalette: [
                'rgb(00,131,255)'
            ],
            selection: null,
            visible: true,
            orderBy: [],
            where: {}
        };
        mockData = transposeMetrics({
            headers: [
                {
                    type: 'attrLabel',
                    id: 'healthdata_finish.aci81lMifn6q',
                    uri: '/gdc/md/zro9kxjp2hejksfug8qemqwx6d92c940/obj/2577',
                    title: 'Quarter/Year (Health Data_finish)'
                },
                {
                    type: 'metric',
                    id: 'c54rxxM5coY6',
                    uri: '/gdc/md/zro9kxjp2hejksfug8qemqwx6d92c940/obj/2647',
                    title: 'Average kilometers',
                    format: '#,##0.00'
                }
            ],
            rawData: []
        });
        mockHeaders = mockData.headers;
    });

    describe('propertiesToHeaders', () => {
        it('converts vis ctrl properties to header format', () => {
            const res = propertiesToHeaders(config, mockHeaders);
            expect(res.x).toEqual({
                type: 'attrLabel',
                id: 'healthdata_finish.aci81lMifn6q',
                uri: '/gdc/md/zro9kxjp2hejksfug8qemqwx6d92c940/obj/2577',
                title: 'Quarter/Year (Health Data_finish)'
            });
            expect(res.y).toEqual({
                id: 'metricValues',
                type: 'metric',
                uri: '/metricValues',
                format: '#,##0.00',
                title: 'Average kilometers'
            });
        });
    });

    describe('getIndices', () => {
        it('return indices from config', () => {
            expect(getIndices(config, mockData.headers))
                .toEqual({ category: 0, series: 1, metric: 2 });
        });
    });

    describe('isMetricNamesInSeries', () => {
        it('works', () => {
            expect(isMetricNamesInSeries(config, mockHeaders)).toEqual(true);
            config.y = 'metricNames';
            config.color = 'metricValues';
            expect(isMetricNamesInSeries(config, mockHeaders)).toEqual(false);
        });
    });

    describe('getLegendLayout', () => {
        it('works', () => {
            expect(getLegendLayout(config, mockHeaders)).toEqual('horizontal');
            config.y = 'metricNames';
            config.color = 'metricValues';
            expect(getLegendLayout(config, mockHeaders)).toEqual('vertical');
        });
    });

    describe('getCategoryAxisLabel', () => {
        it('works', () => {
            expect(getCategoryAxisLabel(config, mockHeaders))
                .toEqual('Quarter/Year (Health Data_finish)');
            mockHeaders[0].title = undefined;
            expect(getCategoryAxisLabel(config, mockHeaders)).toEqual('');
        });
    });

    describe('getMetricAxisLabel', () => {
        it('works', () => {
            expect(getMetricAxisLabel(config, mockHeaders)).toEqual('Average kilometers');
            mockHeaders[1].metrics = [];
            expect(getMetricAxisLabel(config, mockHeaders)).toEqual('Average kilometers');
        });
    });

    describe('showInPercent', () => {
        it('works', () => {
            expect(showInPercent(config, mockHeaders)).toEqual(false);
            mockHeaders[2].format = '#,##0.00 %';
            expect(showInPercent(config, mockHeaders)).toEqual(true);
        });
    });

    describe('generateTooltipFn', () => {
        const tooltipFnOptions = { categoryAxisLabel: 'category-label' };

        describe('unescaping angle brackets and htmlescaping the whole value', () => {
            const generatedTooltip = generateTooltipFn(tooltipFnOptions);

            it('should keep &lt; and &gt; untouched (unescape -> escape)', () => {
                const tooltip = generatedTooltip({
                    y: 1,
                    series: {
                        name: '&lt;series&gt;'
                    }
                });

                expect(tooltip.includes('&lt;series&gt;')).toEqual(true);
            });

            it('should escape other html chars and have output properly escaped', () => {
                const tooltip = generatedTooltip({
                    y: 1,
                    series: {
                        name: '"&\'&lt;'
                    }
                });

                expect(tooltip.includes('&quot;&amp;&#39;&lt;')).toEqual(true);
            });

            it('should unescape brackets and htmlescape category', () => {
                const tooltip = generatedTooltip({
                    y: 1,
                    category: '&gt;"&\'&lt;',
                    series: {
                        name: 'series'
                    }
                });

                expect(tooltip.includes('&gt;&quot;&amp;&#39;&lt;')).toEqual(true);
            });
        });
    });

    describe('generatePieTooltipFn', () => {
        describe('unescaping angle brackets and htmlescaping the whole value', () => {
            const tooltipFnOptions = { categoryLabel: 'category-label', metricLabel: 'opportunities lost', metricsOnly: false };
            const generatedTooltip = generatePieTooltipFn(tooltipFnOptions);

            it('should keep &lt; and &gt; untouched (unescape -> escape)', () => {
                const tooltip = generatedTooltip({
                    y: 1,
                    name: '&lt;series&gt;'
                });
                expect(tooltip.includes('&lt;series&gt;')).toEqual(true);
            });

            it('should escape other html chars and have output properly escaped', () => {
                const tooltip = generatedTooltip({
                    y: 1,
                    name: '"&\'&lt;'
                });
                expect(tooltip.includes('&quot;&amp;&#39;&lt;')).toEqual(true);
            });

            it('should unescape brackets and htmlescape category', () => {
                const tooltip = generatedTooltip({
                    y: 1,
                    name: '&gt;"&\'&lt;'
                });

                expect(tooltip.includes('&gt;&quot;&amp;&#39;&lt;')).toEqual(true);
            });
        });

        describe('tooltip renders correctly according to attr/metric or multiple metrics', () => {
            it('renders correctly with attribute and metric', () => {
                const tooltipFnOptions = { categoryLabel: 'category-label', metricLabel: 'opportunities lost', metricsOnly: false };
                const generatedTooltip = generatePieTooltipFn(tooltipFnOptions);

                const tooltip = generatedTooltip({
                    y: 1,
                    name: '"&\'&lt;'
                });

                expect(tooltip.includes('category-label')).toEqual(true);
            });

            it('renders correctly with metrics only', () => {
                const tooltipFnOptions = { categoryLabel: 'category-label', metricLabel: 'opportunities lost', metricsOnly: true };
                const generatedTooltip = generatePieTooltipFn(tooltipFnOptions);

                const tooltip = generatedTooltip({
                    y: 1,
                    name: 'opportunities'
                });

                expect(tooltip.includes('opportunities')).toEqual(true);
                expect(tooltip.includes('category-label')).toEqual(false);
            });
        });
    });

    describe('getPieFamilyChartData', () => {
        const oneMetricOneAttributeData = {
            isLoaded: true,
            headers: [
                {
                    type: 'attrLabel',
                    id: 'date.aag81lMifn6q',
                    uri: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/15331',
                    title: 'Year (Date)'
                }, {
                    type: 'metric',
                    id: 'metric_yowwuctu6c5lkxql3itj3nz4ec54ax89_16206.generated.pop.5b24b8',
                    uri: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/808882',
                    title: 'Email Click Rate - previous year',
                    format: '#,##0.0%'
                }
            ],
            rawData: [
                [{ id: '2013', name: '2013' }, '124'],
                [{ id: '2014', name: '2014' }, '284'],
                [{ id: '2015', name: '2015' }, '123'],
                [{ id: '2016', name: '2016' }, '155'],
                [{ id: '2017', name: '2017' }, null]
            ],
            isEmpty: false,
            isLoading: false
        };

        it('should be able to rotate color palette', () => {
            const pieConfig = {
                metricsOnly: false,
                colorPalette: DEFAULT_COLOR_PALETTE.slice(0, 2)
            };

            const data = getPieFamilyChartData(pieConfig, oneMetricOneAttributeData);

            expect(data).toEqual({
                series: [{
                    data: [
                        { name: '2014', y: 284, color: DEFAULT_COLOR_PALETTE[0], legendIndex: 0, format: '#,##0.0%' },
                        { name: '2016', y: 155, color: DEFAULT_COLOR_PALETTE[1], legendIndex: 1, format: '#,##0.0%' },
                        { name: '2013', y: 124, color: DEFAULT_COLOR_PALETTE[0], legendIndex: 2, format: '#,##0.0%' },
                        { name: '2015', y: 123, color: DEFAULT_COLOR_PALETTE[1], legendIndex: 3, format: '#,##0.0%' },
                        { name: '2017', y: null, color: DEFAULT_COLOR_PALETTE[0], legendIndex: 4, format: '#,##0.0%' }
                    ]
                }]
            });
        });

        it('should prepare data correctly when provided with one metric and one attribute', () => {
            const pieConfig = {
                metricsOnly: false,
                colorPalette: DEFAULT_COLOR_PALETTE.slice(0, 2)
            };

            const data = getPieFamilyChartData(pieConfig, oneMetricOneAttributeData);

            expect(data).toEqual({
                series: [{
                    data: [
                        { name: '2014', y: 284, color: DEFAULT_COLOR_PALETTE[0], legendIndex: 0, format: '#,##0.0%' },
                        { name: '2016', y: 155, color: DEFAULT_COLOR_PALETTE[1], legendIndex: 1, format: '#,##0.0%' },
                        { name: '2013', y: 124, color: DEFAULT_COLOR_PALETTE[0], legendIndex: 2, format: '#,##0.0%' },
                        { name: '2015', y: 123, color: DEFAULT_COLOR_PALETTE[1], legendIndex: 3, format: '#,##0.0%' },
                        { name: '2017', y: null, color: DEFAULT_COLOR_PALETTE[0], legendIndex: 4, format: '#,##0.0%' }
                    ]
                }]
            });
        });

        it('should prepare data correctly when provided with multiple metrics', () => {
            const pieConfig = {
                metricsOnly: true,
                colorPalette: DEFAULT_COLOR_PALETTE
            };

            const pieData = {
                isLoaded: true,
                headers: [
                    {
                        type: 'metric',
                        id: 'metric_yowwuctu6c5lkxql3itj3nz4ec54ax89_16206.generated.pop.5b24b8',
                        uri: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/808882',
                        title: 'First',
                        format: '#,##0.0%'
                    },
                    {
                        type: 'metric',
                        id: 'metric_yowwuctu6c5lkxql3itj3nz4ec54ax89_16206.generated.pop.5b24b8',
                        uri: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/808882',
                        title: 'Next one',
                        format: '#,##0.0%'
                    },
                    {
                        type: 'metric',
                        id: 'metric_yowwuctu6c5lkxql3itj3nz4ec54ax89_16206.generated.pop.5b24b8',
                        uri: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/808882',
                        title: 'Next two',
                        format: '#,##0.0%'
                    }
                ],
                rawData: [
                    ['123', '456', '789']
                ],
                isEmpty: false,
                isLoading: false
            };

            const data = getPieFamilyChartData(pieConfig, pieData);

            expect(data).toEqual({
                series: [{
                    data: [
                        { name: 'Next two', y: 789, color: DEFAULT_COLOR_PALETTE[0], legendIndex: 0, format: '#,##0.0%' },
                        { name: 'Next one', y: 456, color: DEFAULT_COLOR_PALETTE[1], legendIndex: 1, format: '#,##0.0%' },
                        { name: 'First', y: 123, color: DEFAULT_COLOR_PALETTE[2], legendIndex: 2, format: '#,##0.0%' }
                    ]
                }]
            });
        });
    });

    describe('getLineFamilyChartData', () => {
        beforeEach(() => {
            config = {
                color: '/metricGroup',
                orderBy: [],
                stacking: false,
                type: 'column',
                where: {},
                x: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/15331',
                y: '/metricValues'
            };

            mockData = {
                isLoaded: true,
                headers: [
                    {
                        type: 'attrLabel',
                        id: 'date.aag81lMifn6q',
                        uri: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/15331',
                        title: 'Year (Date)'
                    }, {
                        type: 'metric',
                        id: 'metric_yowwuctu6c5lkxql3itj3nz4ec54ax89_16206.generated.pop.5b24b8',
                        uri: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/808882',
                        title: 'Email Click Rate - previous year',
                        format: '#,##0.0%'
                    }, {
                        type: 'metric',
                        id: 'bHPCwcn7cGns',
                        uri: '/gdc/md/yowwuctu6c5lkxql3itj3nz4ec54ax89/obj/16206',
                        title: 'Email Click Rate',
                        format: '#,##0.0%'
                    }
                ],
                rawData: [
                    [{ id: '2013', name: '2013' }, null, '0.0126946885814334'],
                    [{ id: '2014', name: '2014' }, '0.0126946885814334', '0.0261557203220383'],
                    [{ id: '2015', name: '2015' }, '0.0261557203220383', '0.0348732552824948'],
                    [{ id: '2016', name: '2016' }, '0.0348732552824948', null]
                ],
                isEmpty: false,
                isLoading: false
            };
        });

        it('should get chart data', () => {
            const chartData = getLineFamilyChartData(config, mockData);

            expect(chartData).toBeDefined();
        });

        it('should not sort data', () => {
            const chartData = getLineFamilyChartData(config, mockData);

            expect(chartData.series[0].name).toEqual('Email Click Rate - previous year');
            expect(chartData.series[1].name).toEqual('Email Click Rate');
        });
    });

    describe('getPieChartOptions', () => {
        describe('metricOnly', () => {
            it('should be true for multiple metrics', () => {
                const data = {
                    headers: [
                        {
                            title: 'm1',
                            type: 'metric'
                        },
                        {
                            title: 'm2',
                            type: 'metric'
                        }
                    ]
                };
                expect(getPieChartOptions({}, data).metricsOnly).toEqual(true);
            });

            it('should be true for single metric', () => {
                const data = {
                    headers: [
                        {
                            title: 'm1',
                            type: 'metric'
                        }
                    ]
                };
                expect(getPieChartOptions({}, data).metricsOnly).toEqual(true);
            });

            it('should be false for one metric and one attribute', () => {
                const data = {
                    headers: [
                        {
                            title: 'm1',
                            type: 'metric'
                        }, {
                            title: 'att1',
                            type: 'attrLabel'
                        }
                    ]
                };
                expect(getPieChartOptions({}, data).metricsOnly).toEqual(false);
            });
        });

        describe('tooltip', () => {
            it('should handle tooltip generation if no metric present', () => {
                const data = {
                    headers: [
                        {
                            title: 'att1',
                            type: 'attrLabel'
                        }
                    ]
                };

                const tooltip = getPieChartOptions({}, data).actions.tooltip({
                    y: 1234,
                    format: '##'
                });

                expect(tooltip).toMatchSnapshot();
            });
        });
    });
});
