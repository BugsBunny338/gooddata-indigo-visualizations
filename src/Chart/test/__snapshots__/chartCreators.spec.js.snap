// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chartCreators getPieChartOptions tooltip should handle tooltip generation if no metric present 1`] = `
"
            <table class=\\"tt-values\\">
                <tr>
                <td class=\\"title\\">att1</td>
                <td class=\\"value\\"></td>
            </tr>
                <tr>
                    <td class=\\"title\\"></td>
                    <td class=\\"value\\">1234</td>
                </tr>
            </table>"
`;
